crates/ruma-events/src/matrix_rtc/member_data.rs

```diff
@@ -0,0 +1,311 @@
+#[derive(Clone, Debug, Serialize, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+#[cfg_attr(not(feature = "unstable-exhaustive-types"), non_exhaustive)]
+pub struct LegacyMembershipData {
+    /// The type of the matrixRTC session the membership belongs to.
+    ///
+    /// e.g. call, spacial, document...
+    #[serde(flatten)]
+    pub application: Application,
+
+    /// The device id of this membership.
+    ///
+    /// The same user can join with their phone/computer.
+    pub device_id: String,
+
+    /// The duration in milliseconds relative to the time this membership joined
+    /// during which the membership is valid.
+    ///
+    /// The time a member has joined is defined as:
+    /// `MIN(content.created_ts, event.origin_server_ts)`
+    #[serde(with = "ruma_common::serde::duration::ms")]
+    pub expires: Duration,
+
+    /// Stores a copy of the `origin_server_ts` of the initial session event.
+    ///
+    /// If the membership is updated this field will be used to track to
+    /// original `origin_server_ts`.
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub created_ts: Option<MilliSecondsSinceUnixEpoch>,
+
+    /// A list of the foci in use for this membership.
+    pub foci_active: Vec<Focus>,
+
+    /// The id of the membership.
+    ///
+    /// This is required to guarantee uniqueness of the event.
+    /// Sending the same state event twice to synapse makes the HS drop the second one and return
+    /// 200.
+    #[serde(rename = "membershipID")]
+    pub membership_id: String,
+}
+
+impl LegacyMembershipData {
+    /// Checks if the event is expired.
+    ///
+    /// Defaults to using `created_ts` of the [`LegacyMembershipData`].
+    /// If no `origin_server_ts` is provided and the event does not contain `created_ts`
+    /// the event will be considered as not expired.
+    /// In this case, a warning will be logged.
+    ///
+    /// # Arguments
+    ///
+    /// * `origin_server_ts` - a fallback if [`LegacyMembershipData::created_ts`] is not present
+    pub fn is_expired(&self, origin_server_ts: Option<MilliSecondsSinceUnixEpoch>) -> bool {
+        let ev_created_ts = self.created_ts.or(origin_server_ts);
+
+        if let Some(ev_created_ts) = ev_created_ts {
+            let now = MilliSecondsSinceUnixEpoch::now().to_system_time();
+            let expire_ts = ev_created_ts.to_system_time().map(|t| t + self.expires);
+            now > expire_ts
+        } else {
+            // This should not be reached since we only allow events that have copied over
+            // the origin server ts. `set_created_ts_if_none`
+            warn!("Encountered a Call Member state event where the origin_ts (or origin_server_ts) could not be found.\
+            It is treated as a non expired event but this might be wrong.");
+            false
+        }
+    }
+}
+
+/// Initial set of fields of [`LegacyMembershipData`].
+#[derive(Debug)]
+#[allow(clippy::exhaustive_structs)]
+pub struct MembershipInit {
```

Comment:
The name of an initialization struct is `{type}Init` so it should be renamed to `LegacyMembershipDataInit`.